addons/l10n_in/models/company.py

```diff
@@ -25,3 +48,23 @@ def _update_l10n_in_fiscal_position(self):
             ChartTemplate = self.env['account.chart.template'].with_company(company)
             fiscal_position_data = ChartTemplate._get_in_account_fiscal_position()
             ChartTemplate._load_data({'account.fiscal.position': fiscal_position_data})
+
+    @api.onchange('vat')
+    def onchange_vat(self):
+        if self.vat and self.check_vat_in(self.vat):
+            state_id = self.env['res.country.state'].search([('l10n_in_tin', '=', self.vat[:2])], limit=1)
+            if state_id:
+                self.state_id = state_id
+
+    def check_vat_in(self, vat):
+        # reference from https://www.gstzen.in/a/format-of-a-gst-number-gstin.html
+        if vat and len(vat) == 15:
+            all_gstin_re = [
+                r'[0-9]{2}[a-zA-Z]{5}[0-9]{4}[a-zA-Z]{1}[1-9A-Za-z]{1}[Zz1-9A-Ja-j]{1}[0-9a-zA-Z]{1}',      # Normal, Composite, Casual GSTIN
+                r'[0-9]{4}[A-Z]{3}[0-9]{5}[UO]{1}[N][A-Z0-9]{1}',       # UN/ON Body GSTIN
+                r'[0-9]{4}[a-zA-Z]{3}[0-9]{5}[N][R][0-9a-zA-Z]{1}',         # NRI GSTIN
+                r'[0-9]{2}[a-zA-Z]{4}[a-zA-Z0-9]{1}[0-9]{4}[a-zA-Z]{1}[1-9A-Za-z]{1}[DK]{1}[0-9a-zA-Z]{1}',         # TDS GSTIN
+                r'[0-9]{2}[a-zA-Z]{5}[0-9]{4}[a-zA-Z]{1}[1-9A-Za-z]{1}[C]{1}[0-9a-zA-Z]{1}'         # TCS GSTIN
```

Comment:
We usually put two spaces before comment on a line, if you'd prefer, you can also align all the comments so they are easily readable. Please either use the same amount of spaces everywhere or align them :+1: